### PR TITLE
Replace direct manipulation of UnityWebRequest with Uxios

### PR DESCRIPTION
**Pull request by mvriel**

### CODE REVIEW

File: Assets/Scripts/Layers/Adapters/DataTypeChain.cs [hunk 1]
```diff
@@ -94,22 +92,28 @@ private IEnumerator DownloadAndCheckSupport(string url)
...
-            string path = Path.Combine(Application.persistentDataPath, guidFilename);
-
-            uwr.downloadHandler = new DownloadHandlerFile(path);
-            yield return uwr.SendWebRequest();
-            if (uwr.result == UnityWebRequest.Result.Success)
-            {
-                urlAndData.LocalFilePath = path;
-            }
-            else
+            var futureFileInfo = Resource<FileInfo>.At(urlAndData.SourceUrl).Value;
```
Comment (mvriel) [inline Assets/Scripts/Layers/Adapters/DataTypeChain.cs:95]: In case anyone is wondering where the whole 'Guid' and path.Combine stuff went: the default behaviour of Uxios' Resource object -when using the type FileInfo as generic- is to do precisely this, generate a filename based on a GUID, put it in the persistentDataPath and if the original URL had an extension, it will include that.

This behaviour of Uxios was inspired by this method, as this is a really nice way to prevent name clashes; and by returning FileInfo, a framework can use this as an upload mechanism and use FileInfo's native methods to move a file to a definitive location if need. In this case: we don't need that

File: Assets/Scripts/Layers/Adapters/DataTypeChain.cs [hunk 2]
```diff
@@ -94,22 +92,28 @@ private IEnumerator DownloadAndCheckSupport(string url)
...
+            
+            // We want to use and manipulate urlAndData, so we 'curry' it by wrapping a method call in a lambda 
+            futureFileInfo.Then(info => DownloadSucceeded(urlAndData, info));
+            futureFileInfo.Catch(error => DownloadFailed(urlAndData, error));
+            
+            yield return Uxios.WaitForRequest(futureFileInfo);
```
Comment (mvriel) [inline Assets/Scripts/Layers/Adapters/DataTypeChain.cs:101]: Usually, you need not wait for a Promise to complete; the only reason this is here is because this IEnumerator is used in a chain of IEnumerators starting in another coroutine, and continuing after this one. Without refactoring that further, a yield instruction here is necessary. (Just as you do not need to put an Uxios call in a coroutine; Uxios handles the async stuff all by itself)

Comment (TomSimons) [reply Assets/Scripts/Layers/Adapters/DataTypeChain.cs:101]: will putting yield return null; here work too then? just for my curiosity

**Review by TomSimons — APPROVED** [review]
nice

File: Assets/Scripts/Layers/Adapters/DataTypeChain.cs (hunk 2 above)
Comment (mvriel) [reply Assets/Scripts/Layers/Adapters/DataTypeChain.cs:101]: Yield return null will skip 1 frame, but we need to wait for the async process that is in the promise to complete (I wanted to call the method 'WaitForTheFuture' because another name for promises is 'Futures', but that might have been too much tongue in cheek ;) ).

What the WaitForRequest yield instruction does is similar to a `yield return null` every frame until the Promise has reached state "Resolved" (i.e. Done). It is similar to what you would do with UnityWebRequest when instead of yielding SendRequest that you enter a loop and `yield return null` until the UnityWebRequest returns a completed state.

So in short: 1 `yield return null` wouldn't work because you need to `yield return null` until the Promise is done; and that is what this does :)

**Comment by mvriel** [comment]
Merged it - it is approved and I need the QueryString class of Uxios in another PR :D
